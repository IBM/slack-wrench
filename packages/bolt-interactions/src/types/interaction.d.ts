import { App, SlackAction, SlackActionMiddlewareArgs } from '@slack/bolt';

import { InteractionFlow } from '../interaction-flow';

declare namespace Interaction {
  interface Controller<FlowState> {
    (flow: InteractionFlow<FlowState>, app: App): void;
  }

  type FlowIds = Record<string, string>;

  interface FlowContext<FlowState> {
    setState: (state: any, expiresAt?: number) => Promise<unknown>;
    state: FlowState;
    endFlow: () => Promise<unknown>;
    interactionIds: FlowIds;
  }

  interface FlowMiddlewareArgs<FlowState> {
    context: Record<string, string> & FlowContext<FlowState>;
  }

  interface FlowActionMiddlewareArgs<
    FlowState,
    ActionType extends SlackAction = SlackAction
  >
    extends FlowMiddlewareArgs<FlowState>,
      SlackActionMiddlewareArgs<ActionType> {}

  interface ActionConstraints {
    action_id: string;
    block_id?: string | RegExp;
    callback_id?: string | RegExp;
  }
}
