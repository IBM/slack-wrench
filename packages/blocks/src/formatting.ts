// https://api.slack.com/reference/surfaces/formatting#date-formatting

export const DateString = (
  timestamp: number,
  format: string,
  fallback: string,
): string => `<!date^${timestamp}^${format}|${fallback}>`;

export const User = (id: string): string => `<@${id}>`;
