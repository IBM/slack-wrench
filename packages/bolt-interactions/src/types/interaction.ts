import {
  App,
  SlackAction,
  SlackActionMiddlewareArgs,
  SlackViewAction,
  SlackViewMiddlewareArgs,
} from '@slack/bolt';

import { InteractionFlow } from '../interaction-flow';

export interface Controller<FlowState> {
  (flow: InteractionFlow<FlowState>, app: App): void;
}

export type FlowIds = Record<string, string>;

export interface FlowContext<FlowState> {
  setState: (state: any, expiresAt?: number) => Promise<unknown>;
  state: FlowState;
  endFlow: () => Promise<unknown>;
  interactionIds: FlowIds;
}

export interface FlowMiddlewareArgs<FlowState> {
  context: Record<string, string> & FlowContext<FlowState>;
}

export interface FlowActionMiddlewareArgs<
  FlowState,
  ActionType extends SlackAction = SlackAction
>
  extends FlowMiddlewareArgs<FlowState>,
    SlackActionMiddlewareArgs<ActionType> {}

export interface FlowViewMiddlewareArgs<
  FlowState,
  ViewActionType extends SlackViewAction = SlackViewAction
>
  extends FlowMiddlewareArgs<FlowState>,
    SlackViewMiddlewareArgs<ViewActionType> {}

export interface ActionConstraints {
  action_id: string;
  block_id?: string | RegExp;
  callback_id?: string | RegExp;
}
