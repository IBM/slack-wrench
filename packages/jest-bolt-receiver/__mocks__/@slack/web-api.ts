import mockWebApi from '@slack-wrench/jest-mock-web-client';

module.exports = mockWebApi(jest);
