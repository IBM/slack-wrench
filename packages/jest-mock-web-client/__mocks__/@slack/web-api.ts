import mockWebApi from '../../src/index';

module.exports = mockWebApi(jest);
