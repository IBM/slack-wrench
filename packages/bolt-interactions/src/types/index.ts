import * as Interaction from './interaction';

// eslint-disable-next-line import/prefer-default-export
export { Interaction };
