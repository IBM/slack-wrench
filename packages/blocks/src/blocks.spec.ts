import { KnownBlock } from '@slack/types';
import { chain } from 'ramda';

import {
  Actions,
  Blocks,
  Button,
  Context,
  Divider,
  Markdown,
  MdSection,
  PlainText,
  Section,
} from './blocks';
import { DateString } from './formatting';

describe('Slack Block widgets', () => {
  const text = 'Ravenclaw is obviously the best house.';
  const buttonId = 'ravenAgree';

  it('renders a divider', () => {
    expect.assertions(1);
    expect(Divider()).toMatchSnapshot();
  });

  it('renders markdown text', () => {
    expect.assertions(1);
    expect(Markdown(text)).toMatchSnapshot();
  });

  it('renders a plaintext block', () => {
    expect.assertions(1);
    expect(PlainText(text)).toMatchSnapshot();
  });

  it('renders a plaintext block without emoji', () => {
    expect.assertions(1);
    expect(PlainText(text, false)).toMatchSnapshot();
  });

  it('renders a section', () => {
    expect.assertions(1);
    expect(
      Section({
        text: PlainText(text),
        fields: [PlainText(text), Markdown(text)],
        accessory: Button(text, buttonId),
      }),
    ).toMatchSnapshot();
  });

  it('renders a markdown section', () => {
    expect.assertions(1);
    expect(MdSection(text)).toMatchSnapshot();
  });

  it('renders an actions', () => {
    expect.assertions(1);
    expect(
      Actions([Button(text, buttonId), Button(text, `${buttonId}-2`)]),
    ).toMatchSnapshot();
  });

  it('renders a button', () => {
    expect.assertions(1);
    expect(Button(text, buttonId)).toMatchSnapshot();
  });

  it('renders context', () => {
    expect.assertions(1);
    expect(Context([PlainText(text)])).toMatchSnapshot();
  });

  it('filters null blocks', () => {
    expect.assertions(1);
    expect(Blocks([MdSection(text), null, MdSection(text)]).length).toBe(2);
  });

  it('can compose blocks into bigger components', () => {
    expect.assertions(1);

    const channels = [
      {
        name: 'house-ravenclaw',
        purpose: 'Discuss Ravenclaw business',
        lastPost: 1575643433,
        memberCount: 120,
      },
      {
        name: 'house-hufflepuff',
        purpose: 'Discuss Hufflepuff puffery',
        memberCount: 107,
        lastPost: 1575643433,
      },
    ];

    expect(
      Blocks([
        MdSection(`Let me help you find some channels.`, {
          accessory: Button('Search', 'changeSearch'),
        }),
        Divider(),
        MdSection('*Channels*'),
        ...chain(
          (channel: any): KnownBlock[] => [
            MdSection(`*${channel.name}*\n${channel.purpose}`),
            Context([
              Markdown(
                `${channel.memberCount} members\n` +
                  `Last post: ${DateString(
                    channel.lastPost,
                    '{date_pretty}',
                    channel.lastPost.toString(),
                  )}`,
              ),
            ]),
            Actions([
              Button(':thumbsup:', 'thumbsUp', {
                value: channel.name,
              }),
              Button(':thumbsdown:', 'thumbsDown', {
                value: channel.name,
              }),
            ]),
          ],
          channels,
        ),
      ]),
    ).toMatchSnapshot();
  });
});
