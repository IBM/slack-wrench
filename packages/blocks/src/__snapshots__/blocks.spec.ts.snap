// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slack Block widgets can compose blocks into bigger components 1`] = `
Array [
  Object {
    "accessory": Object {
      "action_id": "changeSearch",
      "text": Object {
        "emoji": true,
        "text": "Search",
        "type": "plain_text",
      },
      "type": "button",
    },
    "text": Object {
      "text": "Let me help you find some channels.",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  Object {
    "type": "divider",
  },
  Object {
    "text": Object {
      "text": "*Channels*",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  Object {
    "text": Object {
      "text": "*house-ravenclaw*
Discuss Ravenclaw business",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  Object {
    "elements": Array [
      Object {
        "text": "120 members
Last post: <!date^1575643433^{date_pretty}|1575643433>",
        "type": "mrkdwn",
      },
    ],
    "type": "context",
  },
  Object {
    "elements": Array [
      Object {
        "action_id": "thumbsUp",
        "text": Object {
          "emoji": true,
          "text": ":thumbsup:",
          "type": "plain_text",
        },
        "type": "button",
        "value": "house-ravenclaw",
      },
      Object {
        "action_id": "thumbsDown",
        "text": Object {
          "emoji": true,
          "text": ":thumbsdown:",
          "type": "plain_text",
        },
        "type": "button",
        "value": "house-ravenclaw",
      },
    ],
    "type": "actions",
  },
  Object {
    "text": Object {
      "text": "*house-hufflepuff*
Discuss Hufflepuff puffery",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  Object {
    "elements": Array [
      Object {
        "text": "107 members
Last post: <!date^1575643433^{date_pretty}|1575643433>",
        "type": "mrkdwn",
      },
    ],
    "type": "context",
  },
  Object {
    "elements": Array [
      Object {
        "action_id": "thumbsUp",
        "text": Object {
          "emoji": true,
          "text": ":thumbsup:",
          "type": "plain_text",
        },
        "type": "button",
        "value": "house-hufflepuff",
      },
      Object {
        "action_id": "thumbsDown",
        "text": Object {
          "emoji": true,
          "text": ":thumbsdown:",
          "type": "plain_text",
        },
        "type": "button",
        "value": "house-hufflepuff",
      },
    ],
    "type": "actions",
  },
]
`;

exports[`Slack Block widgets renders a divider 1`] = `
Object {
  "type": "divider",
}
`;

exports[`Slack Block widgets renders a fields section 1`] = `
Object {
  "fields": Array [
    Object {
      "text": "Ravenclaw is obviously the best house.",
      "type": "mrkdwn",
    },
    Object {
      "text": "Ravenclaw is obviously the best house.",
      "type": "mrkdwn",
    },
  ],
  "type": "section",
}
`;

exports[`Slack Block widgets renders a markdown section 1`] = `
Object {
  "text": Object {
    "text": "Ravenclaw is obviously the best house.",
    "type": "mrkdwn",
  },
  "type": "section",
}
`;

exports[`Slack Block widgets renders a section 1`] = `
Object {
  "accessory": Object {
    "action_id": "ravenAgree",
    "text": Object {
      "emoji": true,
      "text": "Ravenclaw is obviously the best house.",
      "type": "plain_text",
    },
    "type": "button",
  },
  "fields": Array [
    Object {
      "emoji": true,
      "text": "Ravenclaw is obviously the best house.",
      "type": "plain_text",
    },
    Object {
      "text": "Ravenclaw is obviously the best house.",
      "type": "mrkdwn",
    },
  ],
  "text": Object {
    "emoji": true,
    "text": "Ravenclaw is obviously the best house.",
    "type": "plain_text",
  },
  "type": "section",
}
`;

exports[`Slack Block widgets renders an actions 1`] = `
Object {
  "elements": Array [
    Object {
      "action_id": "ravenAgree",
      "text": Object {
        "emoji": true,
        "text": "Ravenclaw is obviously the best house.",
        "type": "plain_text",
      },
      "type": "button",
    },
    Object {
      "action_id": "ravenAgree-2",
      "text": Object {
        "emoji": true,
        "text": "Ravenclaw is obviously the best house.",
        "type": "plain_text",
      },
      "type": "button",
    },
  ],
  "type": "actions",
}
`;

exports[`Slack Block widgets renders context 1`] = `
Object {
  "elements": Array [
    Object {
      "emoji": true,
      "text": "Ravenclaw is obviously the best house.",
      "type": "plain_text",
    },
  ],
  "type": "context",
}
`;
